tests/eas/misfit.py

```diff
@@ -0,0 +1,400 @@
+
+    This test spawns nr_cpus 100% tasks. The tasks running on bigger-capacity
+    CPUs will finish first, and it is expected of them to instantly pull the
+    tasks running on smaller-capacity CPUs via idle-balance.
+    """
+
+    # How long the tasks will be pinned to their "starting" CPU. Doesn't have
+    # to be long (we just have to ensure they spawn there), so arbitrary value
+    pin_delay = 0.001
+
+    # Somewhat arbitrary delay - long enough to ensure
+    # rq->avg_idle > sysctl_sched_migration_cost
+    idling_delay = 1
+
+    # How long do we allow misfit tasks to stay on LITTLE CPUs when big CPUs
+    # are free
+    allowed_misfit_pct = 10
+
+    @classmethod
+    def _get_wload(cls, test_env):
+        # rt-app json file is assembled by hand because we need per-phase
+        # CPU-affinity, and that's not (yet) feasible in LISA
+        wload = {
+            "global": {
+                "calibration": "__PVALUE__",
+                "default_policy": "SCHED_OTHER",
+                "duration": -1,
+                "logdir" : "__LOGDIR__"
+            },
+            "tasks": {}
+        }
+
+        task_template = {
+            "delay": int(cls.idling_delay * 1000000),
+            "loop": 1,
+            "phases": OrderedDict([("pinned", {}), ("free", {})]),
+            "policy": "SCHED_OTHER"
+        }
+
+        for key in task_template["phases"].iterkeys():
+            task_template["phases"][key]["loop"] = 1
+
+        task_template["phases"]["pinned"]["run"] = int(1e6 * cls.pin_delay)
+        # We're pinning stuff in the first phase, so give it ample time to
+        # clean the pinned logic out of balance_interval
+        task_template["phases"]["free"]["run"] = int(
+            1e6 * 1.1 * cls._get_max_lb_interval(test_env))
+
+        for cpu in xrange(test_env.platform['cpus_count']):
+            task = deepcopy(task_template)
+            task["phases"]["pinned"]["cpus"] = [cpu]
+            wload["tasks"]["misfit_{}".format(cpu)] = task
+
+        wload_path = os.path.join(test_env.res_dir, 'staggered_template.json')
+        with open(wload_path, 'w') as fh:
+            res = json.dumps(wload)
+            # Tokens that represent integers musn't be escaped by quotes.
+            # The reason strings are used is to dump the json more easily.
+            for key in ['__PVALUE__']:
+                res = res.replace('"{}"'.format(key), key)
+            fh.write(res)
+
+        return {
+            'staggered' : {
+                'type' : 'rt-app',
+                'conf' : {
+                    'class' : 'custom',
+                    'json' : wload_path
+                }
+            }
+        }
+
+    def get_idle_time(self, trace, start, end, cpus):
+        """
+        :returns: How long 'cpus' were idle in the [start, end] interval
+        """
+        idle_time = 0
+
+        for cpu in cpus:
+            busy_df = pd.DataFrame(trace.getCPUActiveSignal(cpu), columns=['state'])
```

Comment:
So, trace.getCPUActiveSignal(cpu) is memoized... but here we keep creating DFs...
Maybe you can add this line in an above memoized wrapper to avoid generating again ad again the same DF... but I'm not sure how many times you call this on the same CPU...